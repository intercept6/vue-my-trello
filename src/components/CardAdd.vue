<template>
<form @submit.prevent="addCardToList" :class="classList">
  <label>
    <input type="text"
           v-model="body"
           class="text-input"
           placeholder="Add new card"
           @focusin="startEditing"
           @focusout="finishEditing"
    />
  </label>
  <button type="submit"
          class="add-button"
          v-if="isEditing || bodyExists"
  >Add Card
  </button>
</form>
</template>

<script>
import {computed, ref} from '@vue/composition-api';

export default {
  name: "CardAdd",
  props: {
    listIndex: {
      type: Number,
      required: true,
    },
  },
  setup(props, context) {

    let body = ref('');
    let isEditing = ref(false);

    const bodyExists = computed(() => {
      return body.value.length > 0;
    });

    const classList = computed(() => {
      const classList = ['add-card'];
      if (isEditing.value) {
        classList.push('active');
      }
      if (bodyExists.value) {
        classList.push('addable');
      }

      return classList;
    });
    const $store = context.root.$store;

    const addCardToList = () => {
      $store.dispatch('addCardToList',
        {
          body: body.value,
          listIndex: props.listIndex,
        });
      body.value = '';
    };

    const startEditing = () => {
      isEditing.value = true;
    };

    const finishEditing = () => {
      isEditing.value = false;
    };

    return {
      body,
      isEditing,
      classList,
      bodyExists,
      addCardToList,
      startEditing,
      finishEditing,
    };
  },
};
</script>
