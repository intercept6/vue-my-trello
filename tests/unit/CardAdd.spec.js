import {shallowMount} from '@vue/test-utils';
import CardAdd from '../../src/components/CardAdd';

describe('CardAdd.vue', () => {
  it('編集中の場合、classにactiveが存在する', () => {
    const wrapper = shallowMount(CardAdd, {
      data() {
        return {
          body: '',
          isEditing: true,
        };
      },
      propsData: {
        listIndex: 1,
      },
    });

    expect(wrapper.classes()).toContain('active');
  });


  it('非編集中の場合、classにactiveが存在しない', () => {
    const wrapper = shallowMount(CardAdd, {
      data() {
        return {
          body: '',
          isEditing: false,
        };
      },
      propsData: {
        listIndex: 1,
      },
    });

    expect(wrapper.classes()).not.toContain('active');
  });

  test('スナップショットテスト', () => {
    const wrapper = shallowMount(CardAdd, {
      data() {
        return {
          body: '',
          isEditing: false,
        };
      },
      propsData: {
        listIndex: 1,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
