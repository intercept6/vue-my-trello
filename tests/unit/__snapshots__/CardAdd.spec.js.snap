// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardAdd.vue スナップショットテスト 1`] = `
<form
  class="add-card"
>
  <label>
    <input
      class="text-input"
      placeholder="Add new card"
      type="text"
    />
  </label>
   
  <!---->
</form>
`;
